### PR TITLE
Supports Distributed Transaction

### PR DESCRIPTION
**Pull request by goccy**
As a default, I will support distributed transaction.

### Set before/after callback for commit

```go
octillery.BeforeCommitCallback(func(tx *sql.Tx, writeQueries []*sql.QueryLog) error {
  // write all writeQueries to log (WAL) in application side
  return nil
}))
octillery.AfterCommitCallback(func(tx *sql.Tx) error {}, func(tx *sql.Tx, isCriticalError bool, failureWriteQueries []*sql.QueryLog) error {
  if isCriticalError {
    // distributed transaction error
    // record failureWriteQueries and should execute after.
  }
}))
```

Also, callback function can set at `*sql.Tx ` scope too.

```go
tx, _ := db.Begin()
tx.BeforeCommitCallback(func(writeQueries []*sql.QueryLog) error {
  // write all writeQueries to log (WAL) in application side
  return nil
}))
tx.AfterCommitCallback(func() error {}, func(isCriticalError bool, failureWriteQueries []*sql.QueryLog) error {
  if isCriticalError {
    // distributed transaction error
    // record failureWriteQueries and should execute after.
  }
}))
```

### Exec for multiple databases

You can use `*sql.Tx` to multiple databases.

(tableA and tableB are stored other databases)
```go
tx, _ := db.Begin()
tx.Exec("insert into tableA ...") // (1)
tx.Exec("insert into tableB ...") // (2)

// tx.Commit() commit (1) before it commit (2)
// if (1) and (2) are success too, call AfterCommitSuccessCallback
// if (1) and (2) are failure too, call AfterCommitFailureCallback with NOT CriticalError
// if (1) is failure, instantly call AfterCommitFailureCallback with NOT CriticalError
// if (1) is OK but (2) is failure, call AfterCommitFailureCallback with CriticalError
if err := tx.Commit(); err != nil {
  ...
}
```

### Recover from CriticalError

```go
writeQueryLogs := queryLogsfromWAL()
tx, _ := db.Begin()
for _, writeQueryLog := range writeQueryLogs {
    isCommittedQuery, err := tx.IsAlreadyCommittedQueryLog(writeQueryLog)
    if err != nil {
       ....
    }
    if !isCommittedQuery {
        result, err := tx.ExecWithQueryLog(writeQueryLog)
        ....
    }
}
```

## If disable distributed transaction support

set `distributed_transaction: false` in `databases.yml`

### CODE REVIEW

**Comment by goccy** [comment]
# [Codecov](https://codecov.io/gh/knocknote/octillery/pull/7?src=pr&el=h1) Report
> Merging [#7](https://codecov.io/gh/knocknote/octillery/pull/7?src=pr&el=desc) into [master](https://codecov.io/gh/knocknote/octillery/commit/4fa8b4048d5681873352ec72c6098d74476334c1?src=pr&el=desc) will **decrease** coverage by `0.77%`.
> The diff coverage is `63.12%`.

[Impacted file tree graph](https://codecov.io/gh/knocknote/octillery/pull/7?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master       #7      +/-   ##
==========================================
- Coverage   73.67%   72.89%   -0.78%     
==========================================
  Files          15       15              
  Lines        1717     1738      +21     
==========================================
+ Hits         1265     1267       +2     
- Misses        265      287      +22     
+ Partials      187      184       -3
```


| [Impacted Files](https://codecov.io/gh/knocknote/octillery/pull/7?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [octillery.go](https://codecov.io/gh/knocknote/octillery/pull/7/diff?src=pr&el=tree#diff-b2N0aWxsZXJ5Lmdv) | `63.63% <0%> (-8.78%)` | :arrow_down: |
| [config/config.go](https://codecov.io/gh/knocknote/octillery/pull/7/diff?src=pr&el=tree#diff-Y29uZmlnL2NvbmZpZy5nbw==) | `100% <100%> (ø)` | :arrow_up: |
| [database/sql/db.go](https://codecov.io/gh/knocknote/octillery/pull/7/diff?src=pr&el=tree#diff-ZGF0YWJhc2Uvc3FsL2RiLmdv) | `77.77% <100%> (-0.57%)` | :arrow_down: |
| [database/sql/tx.go](https://codecov.io/gh/knocknote/octillery/pull/7/diff?src=pr&el=tree#diff-ZGF0YWJhc2Uvc3FsL3R4Lmdv) | `70.51% <45.45%> (-1.35%)` | :arrow_down: |
| [connection/connection.go](https://codecov.io/gh/knocknote/octillery/pull/7/diff?src=pr&el=tree#diff-Y29ubmVjdGlvbi9jb25uZWN0aW9uLmdv) | `62.42% <64.1%> (-0.24%)` | :arrow_down: |

------

[Continue to review full report at Codecov](https://codecov.io/gh/knocknote/octillery/pull/7?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/knocknote/octillery/pull/7?src=pr&el=footer). Last update [4fa8b40...838c8cf](https://codecov.io/gh/knocknote/octillery/pull/7?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

File: database/sql/query_log.go [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	if query.QueryType() != sqlparser.Insert {
+		return query, nil
+	}
+
+	insertQuery := query.(*sqlparser.InsertQuery)
+	insertQuery.SetNextSequenceID(log.LastInsertID)
+	t.replaceInsertQueryByQueryLog(log, insertQuery)
```
Comment (kanataxa) [inline database/sql/query_log.go]: Did you forget to handle error?

Comment (goccy) [reply database/sql/query_log.go]: `replaceInsertQueryByQueryLog` always returns `nil` as `error` type.
So, I fixed that interface.

File: connection/connection.go [hunk 2]
```diff
@@ -93,149 +121,262 @@ type DBConnection struct {
...
+	isCriticalError := false
+
+	defer func() {
+		if len(failedWriteQueries) == 0 {
+			if err := c.AfterCommitSuccessCallback(); err != nil {
+				e = err
+			}
+		} else if len(failedWriteQueries) > 0 {
+			if err := c.AfterCommitFailureCallback(isCriticalError, failedWriteQueries); err != nil {
+				e = err
```
Comment (kanataxa) [inline connection/connection.go]: I considered to use `errors.WithStack(err)` instead of `err`

What does this look like?

Comment (goccy) [reply connection/connection.go]: Added stack trace for `e` . Thanks !

**Comment by ktsujichan** [comment]
LGTM:+1:

**Comment by goccy** [comment]
Thanks !!

**Pull request merged**